pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java

```diff
@@ -508,8 +533,14 @@ protected boolean isHandshakeCompleted() {
             if (log.isDebugEnabled()) {
                 log.debug("{} Failed to add lookup-request into pending queue", requestId);
             }
-            future.completeExceptionally(new PulsarClientException.TooManyRequestsException(
-                    "Failed due to too many pending lookup requests"));
+            if (waitingLookupRequestSemaphore.tryAcquire()) {
```

Comment:
nit:

I think you can use a bounded queue instead of two variables - one queue + one semaphore.

if you are using a bounded queue, the logic here can be simplified:

```
if (!waitingLookupRequests.offer(..)) {

     fail(..)
}
```

https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ArrayBlockingQueue.html#offer(E)